src/storage/gc_worker.rs

```diff
@@ -0,0 +1,275 @@
+use super::txn::GC_BATCH_SIZE;
+use super::{Callback, Error, Key, Result};
+use kvproto::kvrpcpb::Context;
+use prometheus::local::{LocalHistogram, LocalIntCounter, LocalIntCounterVec};
+use std::fmt::{self, Display, Formatter};
+use std::sync::{Arc, Mutex};
+use util::worker::{self, Builder, Runnable, ScheduleError, Worker};
+
+pub const GC_MAX_PENDING: usize = 2;
+
+struct GCTask {
+    pub ctx: Context,
+    pub safe_point: u64,
+    pub callback: Callback<()>,
+}
+
+impl Display for GCTask {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        let epoch = match self.ctx.region_epoch.as_ref() {
+            None => "None".to_string(),
+            Some(e) => format!("{{{}, {}}}", e.conf_ver, e.version),
+        };
+        f.debug_struct("GCTask")
+            .field("region", &self.ctx.region_id)
+            .field("epoch", &epoch)
+            .field("safe_point", &self.safe_point)
+            .finish()
+    }
+}
+
+// `GCRunner` is used to perform GC on the engine
+struct GCRunner {
+    engine: Box<Engine>,
+    ratio_threshold: f64,
+
+    stats: StatisticsSummary,
+    empty_range_counter: LocalIntCounter,
+    skipped_counter: LocalIntCounter,
+    duration_histogram: LocalHistogram,
+    gctask_counter: LocalIntCounter,
+    gctask_fail_counter: LocalIntCounter,
+    gc_keys_counter: LocalIntCounterVec,
+}
+
+impl GCRunner {
+    pub fn new(engine: Box<Engine>, ratio_threshold: f64) -> GCRunner {
+        GCRunner {
+            engine,
+            ratio_threshold,
+
+            stats: StatisticsSummary::default(),
+            empty_range_counter: KV_GC_EMPTY_RANGE_COUNTER.local(),
+            skipped_counter: KV_GC_SKIPPED_COUNTER.local(),
+            duration_histogram: GC_DURATION_HISTOGRAM.local(),
+            gctask_counter: GC_GCTASK_COUNTER.local(),
+            gctask_fail_counter: GC_GCTASK_FAIL_COUNTER.local(),
+            gc_keys_counter: GC_KEYS_COUNTER_VEC.local(),
+        }
+    }
+
+    // Scan keys in the region. Returns scanned keys if any, and a key indicating scan progress
+    fn gc_scan_keys(
+        &mut self,
+        ctx: &Context,
+        safe_point: u64,
+        from: Option<Key>,
+    ) -> Result<(Option<Vec<Key>>, Option<Key>)> {
+        let is_range_start_gc = from.is_none();
+
+        let snapshot = self.engine.snapshot(ctx)?;
+        let mut reader = MvccReader::new(
+            snapshot,
+            Some(ScanMode::Forward),
+            !ctx.get_not_fill_cache(),
+            None,
+            None,
+            ctx.get_isolation_level(),
+        );
+        let res = if !reader.need_gc(safe_point, self.ratio_threshold) {
+            self.skipped_counter.inc();
+            Ok((None, None))
+        } else {
+            reader
+                .scan_keys(from, GC_BATCH_SIZE)
+                .map_err(Error::from)
+                .and_then(|(keys, next)| {
+                    if keys.is_empty() {
+                        if is_range_start_gc {
```

Comment:
How about use `from` instead of `is_range_start_gc`?

Reply:
We can't. `from` was moved while calling `reader.scan_keys`